### PR TITLE
feat: replace surveyJs with vanillajs validation, visibility and compute expressions

### PR DESCRIPTION
**Pull request by mswertz**
Checklist

- [x] remove expressions library as dependency
- [x] backend validation incl test
- [x] pet store example
- [x] backend compute
- [x] backend visible (should not save invisible fields, should not validate invisible fields)
- [x] frontend validation
- [x] frontend visible
- [x] docs
- [x] check against examples and document these cases
- [x] provide preview in schema editor when editing validation/visible expressions
- [x] add 'computed' to the schema editor
- [x] fix server side issues for expressions of the form `pets?.some(p => p.name ==='pooky')`
- [x] add test for visible not inserted and been set to null if not visible on update
- [x] formatting
- [x] merge #1896 

also:
- [x] moved descriptions again below the input. We need seperate PR that allows a seperate formLabel next to you normal label in order to provide beter UX when creating questionaire like forms.

### CODE REVIEW

File: apps/schema/src/components/ColumnEditModal.vue [hunk 1]
```diff
@@ -252,17 +277,22 @@ export default {
...
-          table.name === this.column.table ||
-          (table.subclasses && table.subclasses.includes(this.column.table))
-      );
+      const table = deepClone(this.schema.tables.find(
+          (table) =>
+              table.name === this.tableName ||
+              table.name === this.column.table ||
+              (table.subclasses && table.subclasses.includes(this.column.table))));
+      //replace column with current changes
+      const index = table.columns.findIndex(c => c.name == this.column.name || c.name == this.column.oldName)
```
Comment (MaxPostema) [inline apps/schema/src/components/ColumnEditModal.vue:293]: Volgens mij was het in JS altijd handig '===' te gebruiken ipv '=='

File: apps/molgenis-components/src/components/forms/RowEdit.vue [hunk 2]
```diff
@@ -214,18 +214,37 @@ function getColumnError(column, values, tableMetaData) {
...
   } catch (error) {
-    return "Invalid validation expression, reason: " + error.ha;
+    return `Invalid validation expression '${column.validation}', reason: ${error}`;
   }
 }
 
+function executeExpression(expression, values, tableMetaData) {
+  //make sure all columns have keys to prevent reference errors
+  const copy = deepClone(values);
+  tableMetaData.columns.forEach(c => {
```
Comment (MaxPostema) [inline apps/molgenis-components/src/components/forms/RowEdit.vue:239]: Wellicht 'c' hernoemen naar 'column'

File: apps/schema/src/components/Schema.vue [hunk 3]
```diff
@@ -184,7 +184,7 @@ export default {
       this.loading = true;
       request(
         "graphql",
-        "{_session{schemas,roles}_schema{name,tables{name,tableType,inherit,externalSchema,description,semantics,columns{name,table,position,columnType,inherited,key,refSchema,refTable,refLink,refBack,required,readonly,description,semantics,validation,visible} } } }"
+        "{_session{schemas,roles}_schema{name,tables{name,tableType,inherit,externalSchema,description,semantics,columns{id,name,table,position,columnType,inherited,key,refSchema,refTable,refLink,refBack,required,readonly,description,semantics,validation,visible} } } }"
```
Comment (MaxPostema) [inline apps/schema/src/components/Schema.vue:187]: Maybe nice to indent or use a external graphql file here

File: apps/schema/src/components/ColumnEditModal.vue [hunk 4]
```diff
@@ -321,42 +351,45 @@ export default {
     },
   },
   methods: {
-    click() {
+    showModal() {
+      console.log("click")
```
Comment (MaxPostema) [inline apps/schema/src/components/ColumnEditModal.vue:355]: Don't forget to remove the log's

File: apps/schema/src/components/ColumnEditModal.vue (hunk 1 above)
Comment (mswertz) [reply apps/schema/src/components/ColumnEditModal.vue:293]: yeah thx, my problem is switching between java and js ;-)

File: apps/schema/src/components/Schema.vue (hunk 3 above)
Comment (chinook25) [reply apps/schema/src/components/Schema.vue:187]: You can import { gql } from  "graphql-request"; and use that to make the formatting better. I also really like having some extensions for VS-code that assist with gql highlights and so.

File: apps/schema/src/components/ColumnEditModal.vue [hunk 5]
```diff
@@ -1,180 +1,203 @@
 <template>
   <LayoutModal
-    v-if="show"
-    title="Edit column metadata"
-    @close="cancel"
-    :isCloseButtonShown="!isDisabled"
+      v-if="show === true"
+      title="Edit column metadata"
+      :isCloseButtonShown="false"
```
Comment (chinook25) [inline apps/schema/src/components/ColumnEditModal.vue:5]: lot's of incorrect lay-outing. next time someone else saves this file it will all be reverted back :(

File: apps/schema/src/components/TableEditModal.vue [hunk 6]
```diff
@@ -170,27 +169,25 @@ export default {
       }
       this.show = true;
     },
-    close() {
-      this.show = false;
+    apply() {
```
Comment (chinook25) [inline apps/schema/src/components/TableEditModal.vue:172]: apply what? maybe extent the function name a bit to be more informative

File: apps/schema/src/components/TableEditModal.vue [hunk 7]
```diff
@@ -170,27 +169,25 @@ export default {
...
       this.$emit(this.operation, this.table);
+      this.show = false;
     },
     cancel() {
-      //set
+      this.reset();
+      this.show = false;
+    },
+    reset() {
+      //deep copy
```
Comment (chinook25) [inline apps/schema/src/components/TableEditModal.vue:181]: redundant comment

**Review by chinook25 — CHANGES_REQUESTED** [review]

File: apps/molgenis-components/src/components/forms/RowEdit.vue [hunk 8]
```diff
@@ -147,9 +149,16 @@ export default {
...
+            expression,
+            this.internalValues,
+            this.tableMetaData
+          );
         } catch (error) {
-          this.errorPerColumn[columnId] = `Invalid visibility expression`;
+          this.errorPerColumn[
+            columnId
+          ] = `Invalid visibility expression, reason: ${error}`;
+          return true;
```
Comment (chinook25) [inline apps/molgenis-components/src/components/forms/RowEdit.vue:161]: should this also remove an error in case of succes?

**Review by MaxPostema — CHANGES_REQUESTED** [review]
Looks cool, we may wat to gray out the computed field in data editor. It now gives the suggestion you can change the computed value.

File: backend/molgenis-emx2-datamodels/src/main/java/org/molgenis/emx2/datamodels/PetStoreLoader.java [hunk 9]
```diff
@@ -54,14 +51,16 @@ public void load(Schema schema, boolean includeDemoData) {
         table(ORDER)
             .add(column(ORDER_ID).setPkey())
             .add(column("pet").setType(REF).setRefTable(PET))
-            .add(column(QUANTITY).setType(LONG).setValidation("{quantity} >= 1"))
-            .add(column(PRICE).setType(DECIMAL).setValidation("{price} >= 1"))
+            .add(
+                column(QUANTITY)
+                    .setType(LONG)
+                    .setValidation("if(quantity < 1) 'quantity should be >= 1'"))
```
Comment (chinook25) [inline backend/molgenis-emx2-datamodels/src/main/java/org/molgenis/emx2/datamodels/PetStoreLoader.java:57]: this is an invalid validation for type long (since it is actually a string)

**Review by chinook25 — APPROVED** [review]

File: apps/molgenis-components/src/components/forms/RowEdit.vue (hunk 8 above)
Comment (mswertz) [reply apps/molgenis-components/src/components/forms/RowEdit.vue:161]: not necessarily because another validation error might exists. However, otherwise good point. Ideally we would clean all errors on every value change and then reevaluate all expressions. Therefore added this clean to the 'watch' on internalvalues.

File: backend/molgenis-emx2-datamodels/src/main/java/org/molgenis/emx2/datamodels/PetStoreLoader.java (hunk 9 above)
Comment (mswertz) [reply backend/molgenis-emx2-datamodels/src/main/java/org/molgenis/emx2/datamodels/PetStoreLoader.java:57]: > Looks cool, we may want to gray out the computed field in data editor (like name). It now gives the suggestion you can change the computed value.

I thought we had a readonly on line 17. If that doesn't work then something is wrong :-)

Comment (mswertz) [reply backend/molgenis-emx2-datamodels/src/main/java/org/molgenis/emx2/datamodels/PetStoreLoader.java:57]: it strangely enough seems to work?

**Comment by mswertz** [comment]
> Looks cool, we may want to gray out the computed field in data editor (like name). It now gives the suggestion you can change the computed value.

I thought we had a readonly on line 17. If that doesn't work then something is wrong :-)

**Comment by mswertz** [comment]
> > Looks cool, we may want to gray out the computed field in data editor (like name). It now gives the suggestion you can change the computed value.
> 
> I thought we had a readonly on line 17. If that doesn't work then something is wrong :-)

Indeed something was wrong. Client didn't retrieve the 'computed' state

File: apps/nuxt3-ssr/package.json [hunk 10]
```diff
@@ -8,7 +8,6 @@
     "dev": "nuxt dev",
     "generate": "nuxt generate",
     "preview": "nuxt preview",
-    "postinstall": "nuxt prepare",
```
Comment (connoratrug) [inline apps/nuxt3-ssr/package.json:11]: why is this removed as part of the expressions PR

**Review by connoratrug — CHANGES_REQUESTED** [review]

File: apps/nuxt3-ssr/package.json (hunk 10 above)
Comment (mswertz) [reply apps/nuxt3-ssr/package.json:11]: thanks for spotting this, this must be a merge error of some sort. Because I think nuxt3-ssr shouldn't have been touched.

**Review by connoratrug — APPROVED** [review]

**Review by MaxPostema — APPROVED** [review]

**Review by chinook25 — CHANGES_REQUESTED** [review]
I encounter 2 problems in de current version: 
 - The edit / clone / delete buttons in the table view have disappeared (the other views still have them)
 - The computed field is still editable

**Review by chinook25 — APPROVED** [review]
